README.md
# EFA Installer README

The EFA installer will install the userspace libraries and kernel
modules necessary to use the Elastic Fabric Adapter (EFA) on EC2
instances.  Additional software may be required, such as Intel MPI or
the nccl-ofi-plugin for machine learning applications using NCCL.

## Public Documentation

Refer to the [official documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/efa-start.html) for official instructions to use the installer.
- [Supported instance types](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/efa.html#efa-instance-types)
- [Supported operating systems](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/efa.html#efa-os)

## Installer Usage

The `efa_installer.sh` script managing installing operating system
specific packages.  Most software is installed in `/opt/amazon/efa/`,
except for configuration packages that must place files in `/etc` and
the rdma-core package, which is designed to only install in `/usr`.
The following command line arguments are available to
`efa_installer.sh`:

* `--debug-pacakges`: Install the debug packages next to standard
  packages.
* `--uninstall`: Uninstall all packages installed by the EFA
  installer.
* `--yes`: Assume yes to any prompts and run the installer without
  prompting for verification.
* `--no-verify`: Skip running EFA functionality verification tests
  at end of installer.  Useful if building an AMI on an instance
  type which does not support EFA.
* `--skip-limit-conf`: Skip installing configurations to change the
  default limits for locked pages.
* `--skip-kmod`: Do not install the EFA kernel module while
  installing user-space packages.  Useful for installing EFA
  libraries in containers.
* `--enable-gdr`: Enable GPUDirect RDMA support.  This will result
  in the installation of a kernel module which can only be loaded if
  the NVIDIA driver stack is already loaded, so limits the kernel
  module to only working on instance types with NVIDIA GPUs.
* `--minimal`: Only install the kernel module and rdma-core.  Do not
  install Open MPI or Libfabric.

## Installed Packages

* dkms
* rdma-core
* libfabric
* Open MPI
* efa-config
* efa-profile
* efa kernel module

Other system packages may be installed as dependencies of these packages.


ChangeLog.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.24.1] - July 2023
- Upgrade libfabric to 1.18.1
- Upgrade efa driver to 2.5.0

## [1.24.0] - June 2023
- Ingest rdma-core 46.0
- Ingest efa driver 2.4.1
- Support Debian 11

## [1.23.1] - June 2023
- Ingest libfabric 1.18.0amzn2.0

## [1.23.0] - May 2023
- Add support for Debian 10
- Drop support for Ubuntu 18.04 LTS
- Upgrade efa-config package to 1.14
- Ingest libfabric 1.18.0amzn1.0

## [1.22.1] - March 2023
- Upgrade libfabric to 1.17.1

## [1.22.0] - February 2023
- Upgrade Open MPI to 4.1.5
- Upgrade libfabric to 1.17.0
- Upgrade efa-config package to 1.13

## [1.21.0] - December 2022
- Add support for Rocky Linux 9 OS
- Ingest efa driver 2.1.1
- Ingest libfabric 1.16.1amzn3.0
- Upgrade efa-config package to 1.12

## [1.20.0] - November 2022
- Add support for Rocky Linux 8 OS.
- Ingest efa driver 2.1.0.
- Ingest rdma-core 43.0.
- Ingest libfabric 1.16.1amzn1.0

## [1.19.0] - October 2022
- Ingest libfabric 1.16.0
- Build Open MPI with `--enable-orterun-prefix-by-default`

## [1.18.0] - August 2022
- Add support for Ubuntu22.04

## [1.17.3] - August 2022
- Update libfabric to 1.16.0~amzn4.0. The "~" indicates it is a pre-release version of libfabric 1.16.0.
- Extend post-installation pingpong test timeout to 20 seconds

## [1.17.2] - July 2022
- Update libfabric to 1.16.0~amzn3.0. The "~" indicates it is a pre-release version of libfabric 1.16.0.

## [1.17.1] - July 2022
- Update libfabric to 1.16.0~amzn2.0. The "~" indicates it is a pre-release version of libfabric 1.16.0.
- Disable the experimental net provider when building libfabric

## [1.17.0] - July 2022
- Update rdma-core to v41.0
- Update Open MPI to 4.1.4
- Update libfabric to 1.16.0~amzn1.0. The "~" indicates it is a pre-release version of libfabric 1.16.0.

## [1.16.0] - June 2022
- Update libfabric to 1.15.1amzn1.0, contains neuron library name change
- Upgrade efa-config to 1.10
- Exclude opx and rxd providers in the libfabric build

## [1.15.2] - May 2022
- Update libfabric to 1.14.1

## [1.15.1] - March 2022
- Update libfabric to 1.14.0amzn1.0

## [1.15.0] - Feburary 2022
- Fix a bug that cause installation fail on Open SuSE 15.3
- Drop support of Open SuSE 15.2 (as Open SuSE 15.2 reached end of life)
- Drop Support of CentOS 8 (as CentOS 8 reached end of life)
- Update libfabric to 1.14.0
- Update efa kernel driver to 1.16.0
- Update rdma-core to v39.0
- Update Open MPI to version 4.1.2.

## [1.14.1] - October 2021

- Update libfabric to 1.13.2amzn1.0.

## [1.14.0] - October 2021

- Ingest efa kernel driver 1.14.2.
- Make "-g, --enable-gdr" in efa_installer.sh as a no-op option as the latest efa kernel driver enables GDR support by default.
- Ingest rdma-core v37.0.
- Ingest libfabric 1.13.2.
- Add packages list and compare RPM/DEB to list during installation to prevent
  unknown package installations.
- Add sleep in installer script to wait for udev rule to apply after EFA driver reload.

## [1.13.0] - August 2021

- Update rdma-core to v35.0.
- Update libfabric to v1.13.0amzn1.0.
- Add EFA support for CentOS/RHEL 8 on Gravition2 platform.
- Add version comparison logic in installer script to skip the local
  package installation when there is one installed in system with
  higher version.

## [1.12.3] - July 2021

- Update EFA kernel module to 1.13.0.
- Update efa-config package to version 1.9. Improve the calculation of
  huge page reservation to handle large defaulted huge page size.

## [1.12.2] - June 2021

- Update EFA kernel module to 1.12.3.
- Build Open MPI debian packages with `--with-libevent=external`
  and `--with-hwloc=external`.
- Bump Open MPI rpm build ID to 2 to fix backward compatibility
  issue of HWLOC on CentOS 8.
- Remove the installation of kernel-devel and kernel-source
  packages on SLES15SP2 and openSUSE 15.2.

## [1.12.1] - May 2021

- Update Libfabric to version 1.11.2amzon1.1.
- Update EFA kernel module to version 1.12.1.

## [1.12.0] - May 2021

- Update Open MPI to version 4.1.1.
- Update Libfabric to version 1.11.2amzn1.0.
- Build rdma-core for Amazon Linux 2 using the same packaging
  configuration as the AL2 rdma-core.
- Do not force `-Wl,--enable-new-dtags` when building Open MPI RPMs.
- Build Open MPI with system libraries for hwloc and libevent.
- Update EFA kernel module to version 1.12.0
- Update efa-config package to version 1.8.  Improve the calculation
  of huge page reservation for long-lived instances.
- Update efa-profile package to version 1.5. Remove the open mpi collective
  tuning file that worked as a workaround to fix Open MPI 4.1.0 hang on P4d.
- Update rdma-core to v32.1.
- Drop support for Amazon Linux 1 and Ubuntu 16.04.

## [1.11.2] - February 2021

- Fix Open MPI hang when using Open MPI on P4d by changing the default
  algorithm used to implement MPI_BARRIER via a configuration file.
- Disable use of builtin atomics in Open MPI on ARM via
  `--disable-builtin-atomics` to work around compiler issue.

## [1.11.1] - December 2020

- Update Open MPI to version 4.1.0.
- Update efa-config package to version 1.7.  Improve calculation of huge page
  reservation count.
- Update efa-profile package to version 1.3.  Removes unneeded
  collectives decision file now that Open MPI 4.1.0 is used.

## [1.11.0] - December 2020

- Add support for Gravition2 platform.
- Update rdma-core to version 31.2amzn.
- Update Libfabric to version 1.11.1amzn1.0.
- Update efa-config to version 1.6.
- Update efa-profile to version 1.2.

## [1.10.1] - November 2020

- Add support for CentOS / RHEL 8.
- Add support for Ubuntu 20.04.
- Add support for SUSE Linux Enterprise 15.

## [1.10.0] - October 2020

- Add GPUDirect RDMA support for P4d platform.  Use `--enable-gdr`
  installer option to instal GDR-aware kernel module and userspace.
- Update EFA kernel module to version 1.10.2.
- Update rdma-core to version 31.amzn0.
- Update Libfabric to version 1.11.1.
- Update Open MPI to version 4.0.5.

efa_test.sh
#!/bin/bash
# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.

# Script to run fi_pingpong locally over EFA to test connectivity.

if ! command -v fi_info >/dev/null 2>&1 || ! command -v fi_pingpong >/dev/null 2>&1; then
	echo "Error: required libfabric binaries not found."
	exit 1
fi

if ! fi_info -p efa >/dev/null 2>&1; then
	echo "Error: EFA libfabric provider not detected." >&2
	exit 1
fi

echo "Starting server..."
FI_EFA_ENABLE_SHM_TRANSFER=0 fi_pingpong -e rdm -p efa >/dev/null 2>&1 &
sleep 0.5

echo "Starting client..."
FI_EFA_ENABLE_SHM_TRANSFER=0 timeout 20 fi_pingpong -e rdm -p efa localhost
ret=$?
if [ $ret -ne 0 ]; then
	if [ $ret -eq 124 ]; then
		echo "Error: fi_pingpong test failed because of timeout." >&2
	else
		echo "Error: fi_pingpong test failed with error code $ret." >&2
	fi
fi
kill %1
exit $ret


efa_uninstall_limits.sh
#!/bin/bash
# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.

# Removes the old EFA limits configuration files. This is now handled by the
# efa-config package.

. common.sh

EFA_HUGEPAGES_SCRIPT="/opt/amazon/efa/bin/efa_hugepages.sh"

uninstall_limits_conf() {
	efa_limits="/etc/security/limits.d/efa.conf"
	if [ -f ${efa_limits} ]; then
		rm "${efa_limits}"
	fi
}

uninstall_hugepages_rc_local() {
	rc="/etc/rc.local"
	efa_str="# Additional configuration for Amazon EFA"

	if [ ! -f "${rc}" ]; then
		return 0
	fi

	if ! grep -q "${efa_str}" "${rc}"; then
		return 0
	fi

	sed -i "/${efa_str}/,+1d" "${rc}"
	rm -f ${EFA_HUGEPAGES_SCRIPT}
}

generate_grub() {
	if is_amazon_linux_2 || is_centos_7 || is_rhel_7; then
		grub_cfg="/boot/grub2/grub.cfg"
		grub_cmd="grub2-mkconfig"
	else
		echo "Unknown operating system."
		exit 1
	fi

	if [ ! -f "${grub_cfg}" ]; then
		echo "${grub_cfg} not found."
		exit 1
	fi

	${grub_cmd} -o "${grub_cfg}"
	return $?
}

uninstall_hugepages_grub_generate() {
	grub_file="/etc/default/grub"
	efa_str="# Additional configuration for Amazon EFA"

	if [ ! -f "${grub_file}" ]; then
		return 0
	fi

	if ! grep -q "${efa_str}" ${grub_file}; then
		return 0
	fi

	sed -i "/${efa_str}/,+1d" "${grub_file}"

	generate_grub
	return $?
}

is_root

ret=0

if ! uninstall_limits_conf; then
	echo "Error removing limits file."
	ret=1
fi

if is_amazon_linux_2 || is_centos_7 || is_rhel_7; then
	uninstall_hugepages_rc_local
else
	echo "Unknown operating system."
	ret=1
fi

if is_amazon_linux_2 || is_centos_7 || is_rhel_7; then
	uninstall_hugepages_grub_generate
fi

exit $ret


common.sh
# Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.

is_root() {
	if [[ $EUID -ne 0 ]]; then
		echo "This installer must be run as root."
		exit 1
	fi
}

prompt() {
	local prompt=$1
	local response

	printf "%s [y/n]: " "${prompt}"

	read -r response

	response="${response,,}"

	if [ "${response}" != 'y' ] && [ "${response}" != 'yes' ]; then
		echo "Exiting..."
		return 1
	fi
	return 0
}

version_larger_or_equal() {
	version_1=$1
	version_2=$2
	version_smaller=$(printf "$version_1\n$version_2" | sort -V | head -n 1)
	if [ "$version_smaller" = "$version_2" ]; then
		return 0
	else
		return 1
	fi
}

has_substring() {
	string="$1"
	substring="$2"

	# ${string#*$substring} - strip $string of the smallest prefix of pattern `*$substring`
	# Stripping should also work, i.e. ${string%"$substring"*}
	# Copied from https://tiny.amazon.com/rj6h3xs1/stacques2829howd
	# Works for POSIX shell
	if test "${string#*$substring}" != "$string"; then
		return 0 # $substring is in $string
	else
		return 1 # $substring is not in $string
	fi
}

is_amazon_linux_2() {
	. /etc/os-release
	if [ "$NAME" = "Amazon Linux" ] && [ "$VERSION_ID" = "2023" ]; then
		return 0
	else
		return 1
	fi
}

is_centos_7() {
	. /etc/os-release
	if has_substring "$NAME" "CentOS" && [ "$VERSION_ID" = "7" ]; then
		return 0
	else
		return 1
	fi
}

is_centos_8() {
	. /etc/os-release
	if has_substring "$NAME" "CentOS" && [ "$VERSION_ID" = "8" ]; then
		return 0
	else
		return 1
	fi
}

is_debian_10() {
	. /etc/os-release
	if has_substring "$NAME" "Debian" && [ "$VERSION_ID" = "10" ]; then
		return 0
	else
		return 1
	fi
}

is_debian_11() {
	. /etc/os-release
	if has_substring "$NAME" "Debian" && [ "$VERSION_ID" = "11" ]; then
		return 0
	else
		return 1
	fi
}

is_rhel_7() {
	. /etc/os-release
	if has_substring "$NAME" "Red Hat Enterprise Linux" &&
		echo $VERSION_ID | egrep -q '7\.[0-9]'; then
		return 0
	else
		return 1
	fi
}

is_rhel_8() {
	. /etc/os-release
	if has_substring "$NAME" "Red Hat Enterprise Linux" &&
		echo $VERSION_ID | egrep -q '8\.[0-9]'; then
		return 0
	else
		return 1
	fi
}

is_rockylinux_8() {
	if [ "$NAME" = "Rocky Linux" ] &&
		echo $VERSION_ID | egrep -q '8\.[0-9]'; then
		return 0
	else
		return 1
	fi
}

is_rockylinux_9() {
	if [ "$NAME" = "Rocky Linux" ] &&
		echo $VERSION_ID | egrep -q '9\.[0-9]'; then
		return 0
	else
		return 1
	fi
}

is_ubuntu_2004() {
	. /etc/os-release
	if [ "$NAME" = "Ubuntu" ] && [ "$VERSION_ID" = "20.04" ]; then
		return 0
	else
		return 1
	fi
}

is_ubuntu_2204() {
	. /etc/os-release
	if [ "$NAME" = "Ubuntu" ] && [ "$VERSION_ID" = "22.04" ]; then
		return 0
	else
		return 1
	fi
}

is_suse_15() {
	. /etc/os-release
	if [ "$NAME" = "openSUSE Leap" ]; then
		version_larger_or_equal "$VERSION_ID" "15.3"
		return $?
	elif [ "$NAME" = "SLES" ] && echo $VERSION_ID | egrep -q '15\.[0-9]'; then
		return 0
	else
		return 1
	fi
}

install_cmd() {
	if is_amazon_linux_2 || is_centos_7 || is_rhel_7 || is_centos_8 || is_rhel_8 || is_rockylinux_8 || is_rockylinux_9; then
		if [ $1 == "localinstall" ]; then
			shift
			if is_centos_8; then
				# centos8 has reached end of life, and all repos has expired.
				# Meanwhile, localinstall does not need repos either. Therefore
				# disable repos for centos8 only.
				yum -y localinstall --disablerepo appstream,baseos,powertools $@
			else
				yum -y localinstall $@
			fi
		else
			yum -y install $@
		fi
	elif is_suse_15; then
		zypper install -y --no-recommends $@
	elif is_debian_10 || is_debian_11 || is_ubuntu_2004 || is_ubuntu_2204; then
		apt install -y $@
	else
		echo "Error: unsupported operating system, exiting." >&2
		exit 1
	fi
}
search_cmd() {
	if is_amazon_linux_2 || is_centos_7 || is_rhel_7 || is_centos_8 || is_rhel_8 || is_rockylinux_8 || is_rockylinux_9; then
		yum list installed $@
	elif is_suse_15; then
		zypper search --installed-only --match-exact $@
	elif is_debian_10 || is_debian_11 || is_ubuntu_2004 || is_ubuntu_2204; then
		apt list --installed $@
	else
		echo "Error: unsupported operating system, exiting." >&2
		exit 1
	fi
}
remove_cmd() {
	if is_amazon_linux_2 || is_centos_7 || is_rhel_7 || is_centos_8 || is_rhel_8 || is_rockylinux_8 || is_rockylinux_9; then
		yum -y remove $@
	elif is_suse_15; then
		zypper remove -y $@
	elif is_debian_10 || is_debian_11 || is_ubuntu_2004 || is_ubuntu_2204; then
		#The first argument for purge on ubuntu is 1, for all other
		#cases it will be the package name
		ubuntu_remove_purge=$1
		shift
		if [ $ubuntu_remove_purge -eq 1 ]; then
			apt -y purge $@
		else
			apt -y remove $@
		fi
	else
		echo "Error: unsupported operating system, exiting." >&2
		exit 1
	fi
}
# Check if the given kernel is the current running one
is_kernel_current() {
	kernel=$1
	if is_suse_15; then
		kernel_version_check_string=$(uname -r | sed 's/-default//')
	else
		kernel_version_check_string=$(uname -r)
	fi
	if [[ "${kernel}" == "${kernel_version_check_string}" ]]; then
		return 0
	else
		return 1
	fi
}
